src/core/core.layouts.js

```diff
@@ -3,55 +3,188 @@
+	var bottom = filterByPosition(layoutBoxes, 'bottom');
+
+	// Sort boxes by weight. A higher weight is further away from the chart area
+	sortByWeight(left, true);
+	sortByWeight(right);
+	sortByWeight(top, true);
+	sortByWeight(bottom);
+
 	return {
-		top: top,
-		left: left,
-		bottom: bottom,
-		right: right
+		lt: left.concat(top),
+		rb: right.concat(bottom),
+		area: filterByPosition(layoutBoxes, 'chartArea'),
+		v: left.concat(right),
+		h: top.concat(bottom)
 	};
 }
 
-function addSizeByPosition(boxes, size) {
-	helpers.each(boxes, function(box) {
-		size[box.position] += box.isHorizontal() ? box.height : box.width;
+function updateMaxPadding(padding, box) {
+	if (box.getPadding) {
+		var boxPadding = box.getPadding();
+		padding.top = Math.max(padding.top, boxPadding.top);
+		padding.left = Math.max(padding.left, boxPadding.left);
+		padding.bottom = Math.max(padding.bottom, boxPadding.bottom);
+		padding.right = Math.max(padding.right, boxPadding.right);
+	}
+}
+
+
+function addSizeByPosition(outerDims, boxes) {
+	var result = extend({}, outerDims);
+	helpers.each(boxes, function(layout) {
+		result[layout.pos] += layout.horizontal ? layout.box.height : layout.box.width;
 	});
+	return result;
+}
+
+
+function getCombinedMax(maxPadding, outerDims, a, b) {
+	return Math.max(maxPadding[a], outerDims[a]) + Math.max(maxPadding[b], outerDims[b]);
+}
+
+function adjustWithPadding(outerDims, boxCorner, maxPadding) {
+	['top', 'left'].forEach(function(pos) {
+		if (maxPadding[pos] > outerDims[pos]) {
+			boxCorner[pos] += maxPadding[pos] - outerDims[pos];
+			outerDims[pos] = maxPadding[pos];
+		}
+	});
+}
+
+function getMargin(horizontal, outerDims, maxPadding) {
+	return horizontal
+		? {
+			left: Math.max(outerDims.left, maxPadding.left),
+			right: Math.max(outerDims.right, maxPadding.right)
+		}
+		: {
+			top: Math.max(outerDims.top, maxPadding.top),
+			bottom: Math.max(outerDims.bottom, maxPadding.bottom)
+		};
+}
+
+function updateChartArea(chartArea, outerDims, maxPadding) {
+	var newWidth = chartArea.outerWidth - getCombinedMax(maxPadding, outerDims, 'left', 'right');
+	var newHeight = chartArea.outerHeight - getCombinedMax(maxPadding, outerDims, 'top', 'bottom');
+
+	if (newWidth !== chartArea.w || newHeight !== chartArea.h) {
+		chartArea.w = newWidth;
+		chartArea.h = newHeight;
+		return 1;
+	}
+}
+
+function fitBoxes(boxes, chartArea, outerDims) {
+	var maxPadding = chartArea.maxPadding;
+	var refitBoxes = [];
+	var i, ilen, layout, box, refit;
+
+	for (i = 0, ilen = boxes.length; i < ilen; ++i) {
+		layout = boxes[i];
+		box = layout.box;
+
+		box.update(
+			layout.width || chartArea.w,
+			layout.height || chartArea.h,
+			getMargin(layout.horizontal, outerDims, maxPadding)
+		);
+		updateMaxPadding(maxPadding, box);
+		if (updateChartArea(chartArea, outerDims, maxPadding) && refitBoxes.length) {
+			// Dimensions changed and there were non full width boxes before this
+			// -> we have to refit those
+			refit = 1;
```

Comment:
I would set `refit = true` instead to make it clearer this is being used as a boolean

Reply:
done